src/rendering/vcTileRenderer.cpp

```diff
@@ -516,37 +895,39 @@ void vcTileRenderer_Update(vcTileRenderer *pTileRenderer, const double deltaTime
   pTileRenderer->quadTree.metaData.generateTimeMs = udPerfCounterMilliseconds(startTime);
 }
 
-bool vcTileRenderer_CanNodeDraw(vcQuadTreeNode *pNode)
-{
-  return pNode->renderInfo.pTexture != nullptr;
-}
-
 bool vcTileRenderer_DrawNode(vcTileRenderer *pTileRenderer, vcQuadTreeNode *pNode, vcMesh *pMesh, const udDouble4x4 &view)
 {
-  vcTexture *pTexture = pNode->renderInfo.pDrawTexture;
+  vcTexture *pTexture = pNode->renderInfo.colourDrawInfo.pTexture;
   if (pTexture == nullptr)
     pTexture = pTileRenderer->pEmptyTileTexture;
 
-  for (int t = 0; t < TileVertexResolution * TileVertexResolution; ++t)
+  vcTexture *pDemTexture = pNode->renderInfo.demDrawInfo.pTexture;
+  if (pDemTexture == nullptr)
+    pDemTexture = pTileRenderer->pEmptyDemTileTexture;
+
+  for (int t = 0; t < TileVertexControlPointRes * TileVertexControlPointRes; ++t)
   {
     udFloat4 eyeSpaceVertexPosition = udFloat4::create(view * udDouble4::create(pNode->worldBounds[t], 1.0));
     pTileRenderer->presentShader.everyObject.eyePositions[t] = eyeSpaceVertexPosition;
   }
 
-#if VISUALIZE_DEBUG_TILES
-  pTileRenderer->presentShader.everyObject.colour.w = 1.0f;
-  if (!pNode->renderInfo.pTexture)
-  {
-    pTileRenderer->presentShader.everyObject.colour.x = pNode->level / 21.0f;
-    if (!pNode->visible)
-      pTileRenderer->presentShader.everyObject.colour.z = pNode->level / 21.0f;
-  }
-#endif
+  udFloat2 size = pNode->renderInfo.colourDrawInfo.uvEnd - pNode->renderInfo.colourDrawInfo.uvStart;
+  pTileRenderer->presentShader.everyObject.uvOffsetScale = udFloat4::create(pNode->renderInfo.colourDrawInfo.uvStart, size.x, size.y);
+
+  udFloat2 demSize = pNode->renderInfo.demDrawInfo.uvEnd - pNode->renderInfo.demDrawInfo.uvStart;
+  pTileRenderer->presentShader.everyObject.demUVOffsetScale = udFloat4::create(pNode->renderInfo.demDrawInfo.uvStart, demSize.x, demSize.y);
+
+  pTileRenderer->presentShader.everyObject.tileNormal = udFloat4::create(udFloat3::create(pNode->worldNormal), 0.0f);
 
-  udFloat2 size = pNode->renderInfo.uvEnd - pNode->renderInfo.uvStart;
-  pTileRenderer->presentShader.everyObject.uvOffsetScale = udFloat4::create(pNode->renderInfo.uvStart, size.x, size.y);
+  vcShader_BindTexture(pTileRenderer->presentShader.pProgram, pTexture, 0, pTileRenderer->presentShader.uniform_texture);
+
+  // TODO: This is a hack, needs to be fixed for both graphics APIs
```

Comment:
As mentioned in the shader, rename to variables in the HLSL and it all just works :+1: